java/ql/test/query-tests/security/CWE-378/semmle/tests/Test.java

```diff
@@ -0,0 +1,210 @@
+
+    static File safe6() throws IOException {
+        File temp = File.createTempFile("test", "directory");
+        if (temp.delete() && temp.mkdir()) {
+            return temp;
+        } else {
+            throw new RuntimeException("Failed to create directory");
+        }
+    }
+
+    static File safe7() throws IOException {
+        File temp = File.createTempFile("test", "directory");
+        if (!temp.delete() || !temp.mkdir()) {
+            throw new IOException("Can not convert temporary file " + temp + "to directory");
+        } else {
+            return temp;
+        }
+    }
+
+    /**
+     * When `isDirectory` is true, create a temporary directory, else create a temporary file.
+     */
+    static File safe8(boolean isDirectory) throws IOException {
+        File temp = File.createTempFile("test", "directory");
+        if (isDirectory && (!temp.delete() || !temp.mkdir())) {
+            throw new IOException("Can not convert temporary file " + temp + "to directory");
+        } else {
+            return temp;
+        }
+    }
+
+    static File safe9() throws IOException {
+        File temp = File.createTempFile("test", "directory");
+        if (!temp.delete()) {
+            throw new IOException("Could not delete temp file: " + temp.getAbsolutePath());
+        }
+        if (!temp.mkdir()) {
+            throw new IOException("Could not create temp directory: " + temp.getAbsolutePath());
+        }
+        return temp;
+    }
+
+    static File vulnerable2() throws IOException {
+        File temp = File.createTempFile("test", "directory", new File(System.getProperty("java.io.tmpdir")));
+        temp.delete();
+        temp.mkdir();
+        return temp;
+    }
+
+    static File safe10() throws IOException {
+        File temp = File.createTempFile("test", "directory", new File(System.getProperty("user.dir")));
+        temp.delete();
+        temp.mkdir();
+        return temp;
+    }
+
+    static File vulnerable3() throws IOException {
+        File temp = File.createTempFile(
+            "test", 
+            "directory", 
+            new File(System.getProperty("java.io.tmpdir")).getAbsoluteFile()
+        );
+        temp.delete();
+        temp.mkdir();
+        return temp;
+    }
+
+    static File safe11() throws IOException {
+        File temp = null;
+        if (temp == null) {
+            while (true) {
+                temp = File.createTempFile("test", "directory");
+                if (temp.delete() && temp.mkdir()) {
+                    break;
+                }
+            }
+        }
+        return temp;
+    }
+
+    File safe12temp;
+    File safe12() throws IOException {
+        if (safe12temp == null) {
+            while (true) {
+                safe12temp = File.createTempFile("test", "directory");
+                if (safe12temp.delete() && safe12temp.mkdir()) {
+                    break;
+                }
+            }
+        }
+        return safe12temp;
+    }
```

Comment:
These two still false-positive due to https://github.com/github/codeql/issues/8490

@smowton any thought as to how to fix this?

Reply:
As mentioned in #8490 safe11 is an uninteresting case since it's practically certainly a bug when a local variable is constant and controlling a condition. Safe12 is not safe unless this is the only possible write to that field, which we can only know if the field is private and the assignment on line 170 is the only possible write.